src/Host/Client/Impl/Session/RSessionProvider.cs

```diff
@@ -81,7 +82,14 @@ public class RSessionProvider : IRSessionProvider {
                 return;
             }
 
-            foreach (var session in _sessions.Values) {
+            var sessions = GetSessions();
+            var stopHostTasks = sessions.Select(session => session.StopHostAsync());
+            try {
+                Task.WhenAll(stopHostTasks).GetAwaiter().GetResult();
```

Comment:
I ran it under debugger, and `Dispose` is, in fact, called on UI thread when VS is closed.